src/Admin/AdminHelper.php

```diff
@@ -98,19 +98,38 @@ public function getAdmin($code)
      * Note:
      *   This code is ugly, but there is no better way of doing it.
      *
-     * @param object $subject
+     * @param object $deprecatedSubject
      * @param string $elementId
      *
      * @throws \RuntimeException
      * @throws \Exception
      *
      * @return array
      */
-    public function appendFormFieldElement(AdminInterface $admin, $subject, $elementId)
+    public function appendFormFieldElement(AdminInterface $admin, $deprecatedSubject, $elementId) // NEXT_MAJOR remove the $subject param
```

Comment:
If the third parameter is deprecated, please change the method, so that the 4th parameter will become the 3rd parameter. This will help to remove the deprecation when using this method.

Reply:
How do I do this without a BC-break ?

Reply:
You could check if there are 3 or 4 parameters given (and check for the parameter type).

Reply:
@core23 I still don't see how to do.

If someone extends the function, it will be a BC-breack isnt it ?
For exemple, if I change to 
```
appendFormFieldElement(AdminInterface $admin, $elementId)
```
Someone extending the function with 3 parameters will have a BC-break.

Reply:
I think @core23 is suggesting you do this:

`appendFormFieldElement(AdminInterface $admin, $subjectOrElementId, $elementId = null)`, but that's still a BC-break for extending classes. Maybe deprecate the method in favor of a new one?

Reply:
> Maybe deprecate the method in favor of a new one?

This can be a solution, but the `appendFormFieldElement` was kinda perfect since the action was `appendFormFieldElementAction`

Reply:
Maybe move it to `AppendFormFieldElement::__invoke()`? Classes ending with `Helper` usually group things that have nothing to do with each other.

Reply:
Although in this case we have 2 methods from that class used at the same place:

https://github.com/sonata-project/SonataAdminBundle/blob/b35b794f24a878bc726fdf731bf9de9639e42b44/src/Action/AppendFormFieldElementAction.php#L77-L80

Reply:
This would mean duplicating the code, adding 100 lines of code in the invoke, and will create the need to re-write all the tests for this method... I see your point, but this seems to be a lot of works for just a single little win.

IMHO, almost nobody override/call directly the `AdminHelper::appendFormFieldMethod`, so the deprecation has only a low impact.

I can still try if you consider it's important.

Reply:
> This would mean duplicating the code, adding 100 lines of code in the invoke, and will create the need to re-write all the tests for this method

Would it? You can inject the new class in the old one, right?

Reply:
> IMHO, almost nobody override/call directly the AdminHelper::appendFormFieldMethod, so the deprecation has only a low impact.

You wish, but nobody knows the crazy stuff people do. How about we make the new class final and tag with `@internal`? That way we don't have to deal with that kind of thing in the future.

Reply:
Honestly, I don't care about this deprecation and this is not a need for my PR. If it raise so much debate and request, I can remove the deprecation and keep using this useless param as before. I just discovered a useless param and then say "Hey, it would be nice to remove this in the next major" ; now I feel like I should have say nothing...

> > This would mean duplicating the code, adding 100 lines of code in the invoke, and will create the need to re-write all the tests for this method
> 
> Would it? You can inject the new class in the old one, right?

I don't follow you. If I copy paste the code in the `invoke` method, I can call it from the helper. And if I call the helper in the `invoke` method, this is basically what I'm doing actually.

Reply:
> > IMHO, almost nobody override/call directly the AdminHelper::appendFormFieldMethod, so the deprecation has only a low impact.
> 
> You wish, but nobody knows the crazy stuff people do in their projects. How about we make the new class final and tag with `@internal`? That way we don't have to deal with that kind of thing in the future.

The Helper will already be final in 4.0

Reply:
> The Helper will already be final in 4.0

Yeah but the new class can be final and internal right now.

Reply:
> Honestly, I don't care about this deprecation and this is not a need for my PR. If it raise so much debate and request, I can remove the deprecation and keep using this useless param as before. I just discovered a useless param and then say "Hey, it would be nice to remove this in the next major" ; now I feel like I should have say nothing...

That can happen a lot with open source. Your PR is big and with a lot of discussion, it's a pain to review. When that happens to me, I try hard to split the PR in smaller PRs. If you can do that please do, it will be better for everyone including you.

> I don't follow you. If I copy paste the code in the invoke method, I can call it from the helper. And if I call the helper in the invoke method, this is basically what I'm doing actually.

Here is what I'm suggesting:

```
class AdminHelper
{
    public function deprecatedMethod()
    {
         // deprecation checks
         return (new BrandNewFinalClass())($argumentComputedAbove);
    }
}

final class BrandNewFinalClass
{
    public function __invoke()
    {
         // most of the code ends up here
    }
}

```

Reply:
Here is a PR that does something very similar to what I am suggesting: https://github.com/sonata-project/SonataAdminBundle/commit/2cc7f12fed349c71920d3101f46997219243536e

Reply:
> > Honestly, I don't care about this deprecation and this is not a need for my PR. If it raise so much debate and request, I can remove the deprecation and keep using this useless param as before. I just discovered a useless param and then say "Hey, it would be nice to remove this in the next major" ; now I feel like I should have say nothing...
> 
> That can happen a lot with open source. Your PR is big and with a lot of discussion, it's a pain to review. When that happens to me, I try hard to split the PR in smaller PRs. If you can do that please do, it will be better for everyone including you.

Smaller change would have not a lot of sens here. Almost all of this changes are needed in order to have a subject set when using sonata array collection. By making all of them together, I could have tested them directly in my projet in order to debug and have functional test by hand. Making these change one by one would have been the best way to forget something.

I wanted to avoid the same situation I get with https://github.com/sonata-project/SonataAdminBundle/pull/5938 and https://github.com/sonata-project/SonataAdminBundle/pull/6004. 

The only smaller PR I see would be
```
public function appendFormFieldElement(AdminInterface $admin, $subject, $elementId)
```
Instead of 
```
public function appendFormFieldElement(AdminInterface $admin, $deprecatedSubject, $elementId) // NEXT_MAJOR remove the $deprecatedSubject param
```

There would have been no discussion then. This seems so wrong to me to hide the unused param.

> Here is what I'm suggesting:
> 
> ```
> class AdminHelper
> {
>     public function deprecatedMethod()
>     {
>          // deprecation checks
>          return (new BrandNewFinalClass())($argumentComputedAbove);
>     }
> }
> 
> final class BrandNewFinalClass
> {
>     public function __invoke()
>     {
>          // most of the code ends up here
>     }
> }
> ```

IMHO, both `AdminHelper::appendFormFieldElement` and `AdminHelper::addNewInstance` should be private function from `AppendFormFieldElementAction`. But currently this is not possible since the `AppendFormFieldElementAction` need the `AdminHelper`.

Would it be ok If I add the `@deprecated` tag for these methods with `NEXT_MAJOR: Move these method into AppendFormFieldElementAction as private` ?

It will be easier for the user to remove the deprecation, he just have to stop using the method !

Reply:
> Would it be ok If I add the @deprecated tag for these methods 

Yeah

> with NEXT_MAJOR: Move these method into AppendFormFieldElementAction as private ?

I don't think so: actions should be about HTTP: converting requests into parameters for services, and converting domain exceptions into HTTP exceptions. 

Why don't you create a new class instead? What's wrong with that?

Reply:
> > with NEXT_MAJOR: Move these method into AppendFormFieldElementAction as private ?
> 
> I don't think so: actions should be about HTTP: converting requests into parameters for services, and converting domain exceptions into HTTP exceptions.
> 
> Why don't you create a new class instead? What's wrong with that?

Nothing wrong this that, but when I looked at over action, like `RetrieveAutocompleteItemsAction`, `RetrieveFormFieldElementAction`, ... that was not how things was done.

So how would you call this class @greg0ire ? I'll see if I find time to do this, but still, it's asking to add a big refacto in a bugfix PR which already is not small. It's the contrary of the advice you gave me.

> When that happens to me, I try hard to split the PR in smaller PRs. If you can do that please do, it will be better for everyone including you.

It's more about a general feeling. I saw some interesting PR for Sonata who was never merged because their always have change-requests and the author abandoned. The best is the enemy of the good. There will always things to do better in this project... Where's the limit ?

Reply:
@greg0ire I take a second look at the method of the helper.

All of them are only used by the RetrieveFormFieldElementAction and the AppendFormFieldElementAction. If I move one function, I should move them all...

Reply:
> RetrieveAutocompleteItemsAction, RetrieveFormFieldElementAction, ... that was not how things was done.

Yeah well guess what? I'm not convinced these are great examples of what actions should look like. At least I made things better by splitting the `HelperController`, but there is a lot of room for improvement here IMO. Anyway. If you were to move these methods to an action, then you would have to do the dreaded copy/paste you hinted about, and that would mean twice the maintenance on that part of the code.

> So how would you call this class @greg0ire ?

I would call it `AppendFormFieldElement`.

> I'll see if I find time to do this, but still, it's asking to add a big refacto in a bugfix PR which already is not small and is now one month old.

Not sure if you realize but I saw this PR today and thought that I would try to keep the ball rolling, but I would like not to merge something where there's a deprecation that leads to do weird calls like `foo($argumentThatIsDeprecatedButThatShouldBeStillPassed)`, because I think that's not a great experience for developers. If it comes with an extra step after upgrading to the next major, that's even worse.

> I would have prefer to first fix the bug and second make the refacto in another PR.

Well if you can do that, please do… maybe you can fix the bug and introduce the deprecations later, I don't know.

> It's more about a general feeling. I saw some interesting PR for Sonata who was never merged because their always have change-requests and the author abandoned. The best is the enemy of the good. 

As much as it saddens me to refuse some of these PRs, it's for the best really. "The best is the enemy of the good" is a sentence that makes me cringe when I hear it sometimes, it might be great when you have 1 user, but this package alones has 3k installs in the last 24 hours. Let that sink in for a moment. That means if we are not careful, it can very well piss off users that might not even experience the bug you are fixing. Likewise, there are so many contributions that if we don't strive all the time for better code and architecture, we might end up with fat actions, and ungodly horrors like `AbstractAdmin`.

> There will always things to do better in this project... Where's the limit ?

I try not to ask of people that they dive in a rabbit hole when reviewing a PR, asking them to do things "for bonus points". Here however, that's not the case, because I think this is not mergeable as is. To me, either you do more, or you find a way to do less without making anything worse. You asked for my review several times, so here it is. Feel free to ignore it, I will not block this PR, but what I can tell you is every time I get a notification for this one, I feel less and less inclined to invest time in it.

Reply:
> > I would have prefer to first fix the bug and second make the refacto in another PR.
> 
> Well if you can do that, please do… maybe you can fix the bug and introduce the deprecations later, I don't know.

I updated my PR @greg0ire, now it's fixing the bug without introducing the deprecation. I will work on the deprecation in another PR, it will be smaller and easier. 🙂

> Feel free to ignore it, I will not block this PR, but what I can tell you is every time I get a notification for this one, I feel less and less inclined to invest time in it.

I'm sorry you take it that way.